frontend/ui/widget/container/movablecontainer.lua

```diff
@@ -0,0 +1,295 @@
+        bb:addblitFrom(private_bb, self.dimen.x, self.dimen.y, self.dimen.x, self.dimen.y,
+            self.dimen.w, self.dimen.h, self.alpha)
+        private_bb:free()
+    else
+        -- No alpha, just paint
+        self[1]:paintTo(bb, self.dimen.x, self.dimen.y)
+    end
+end
+
+function MovableContainer:_moveBy(dx, dy, restrict_to_screen)
+    logger.dbg("MovableContainer:_moveBy:", dx, dy)
+    if dx and dy then
+        self._moved_offset_x = self._moved_offset_x + Math.round(dx)
+        self._moved_offset_y = self._moved_offset_y + Math.round(dy)
+        if restrict_to_screen then
+            local screen_w, screen_h = Screen:getWidth(), Screen:getHeight()
+            if self._orig_x + self._moved_offset_x < 0 then
+                self._moved_offset_x = - self._orig_x
+            end
+            if self._orig_y + self._moved_offset_y < 0 then
+                self._moved_offset_y = - self._orig_y
+            end
+            if self._orig_x + self._moved_offset_x + self.dimen.w > screen_w then
+                self._moved_offset_x = screen_w - self._orig_x - self.dimen.w
+            end
+            if self._orig_y + self._moved_offset_y + self.dimen.h > screen_h then
+                self._moved_offset_y = screen_h - self._orig_y - self.dimen.h
+            end
+        end
+        -- if not restrict_to_screen, we don't need to check anything:
+        -- we trust gestures' position and distances: if we started with our
+        -- finger on widget, and moved our finger to screen border, a part
+        -- of the widget should always be on the screen.
+    else
+        -- Not-moving Hold can be used to revert to original position
+        if self._moved_offset_x == 0 and self._moved_offset_y == 0 then
+            -- If we hold while already in initial position, take that
+            -- as a wish to toggle between alpha or no-alpha
+            if self.alpha then
+                self.orig_alpha = self.alpha
+                self.alpha = nil
+            else
+                self.alpha = self.orig_alpha or 0.7
+                -- For testing: to visually see how different alpha
+                -- values look: loop thru decreasing alpha values
+                -- self.alpha = self.orig_alpha or 1.0
+                -- if self.alpha > 0.55 then -- below 0.5 are too transparent
+                --     self.alpha = self.alpha - 0.1
+                -- else
+                --     self.alpha = 0.9
+                -- end
+            end
+        end
+        self._moved_offset_x = 0
+        self._moved_offset_y = 0
+    end
+    -- We need to have all widgets in the area between orig and move position
+    -- redraw themselves
+    local orig_dimen = self.dimen:copy() -- dimen before move/paintTo
+    UIManager:setDirty("all", function()
+        local update_region = orig_dimen:combine(self.dimen)
+        logger.dbg("MovableContainer refresh region", update_region)
+        return "ui", update_region
```

Comment:
I haven't tested this on a device but isn't "ui" (in practice generally `UPDATE_MODE_PARTIAL`) a bit slow compared to "fast" (generally `WAVEFORM_MODE_A2`) or am I mistaken in that? We'd want to follow it with something like "ui" or "partial" in that case once the moving is over.

Reply:
Oh, I didn't mention it in the introduction, but this doesn't actually draw anything while moving (I don't really like that on eInk - and it was probably more complicated:) : it only draws the final position, when one releases its finger.
(It's not like page scrolling where I see how live drawing would matter even if I don't like it).
I use `"ui"`, which is what is used in most `InfoMessage&others' :onShow()` (and not `"partial"` as this results in a full REAGL refresh on Kindle).

Reply:
Oh, I see. :-)